nncf/openvino/engine.py

```diff
@@ -87,104 +95,72 @@ def __init__(self,
-        if stats_layout:
-            self._collect_statistics(predictions, stats_layout)
-
-        processed_outputs = pot.engines.utils.process_raw_output(predictions)
-        outputs = {name: processed_outputs[name] for name in self._output_layers}
-        logits = self.postprocess_output(outputs, None)
-
-        if need_metrics_per_sample:
-            self._per_sample_metrics.append(
-                {
-                    'sample_id': len(self._per_sample_metrics),
-                    'metric_name': 'user_metric',
-                    'result': logits
-                }
-            )
+        self.use_original_metric = use_original_metric
 
     def predict(self,
                 stats_layout=None,
                 sampler=None,
                 stat_aliases=None,
                 metric_per_sample=False,
                 print_progress=False):
-        if self._model is None:
-            raise Exception('Model was not set in Engine class')
-
-        if self._validation_fn is not None:
-            indices = None
-            if sampler:
-                indices = getattr(sampler, '_subset_indices')
-            self._metric.avg_value = self._validation_fn(self._model, self._validation_dataset.get_data(indices))
-
-        return super().predict(stats_layout, sampler, stat_aliases, metric_per_sample, print_progress)
+        subset_indices = None
+        if sampler:
+            subset_indices = sorted(getattr(sampler, '_subset_indices'))
+
+        self._metric.avg_value = None
+        is_full_dataset = subset_indices is None or len(subset_indices) == len(self.data_loader)
+        if self._validation_fn and (is_full_dataset or self.use_original_metric):
+            self._metric.avg_value = self._validation_fn(self._model,
+                                                         self._validation_dataset.get_data(subset_indices))
+            if not metric_per_sample and stats_layout is None:
+                metrics = self._metric.avg_value
+                self._reset()
+                return metrics, {}
+
+        if self.use_original_metric and metric_per_sample:
+            self._per_sample_metrics = self.calculate_per_sample_metrics(subset_indices)
+            if stats_layout is None:
+                metrics = self._metric.avg_value
+                metrics = (sorted(self._per_sample_metrics, key=lambda i: i['sample_id']), metrics)
+                self._reset()
+                return metrics, {}
+
+        dataset_wrapper = DatasetWrapper(self._validation_dataset.get_inference_data(subset_indices))
+        return super().predict(stats_layout, dataset_wrapper, stat_aliases, metric_per_sample, print_progress)
+
+    def calculate_per_sample_metrics(self, subset_indices: List[int]):
+        if self.use_original_metric:
+            per_sample_metrics = []
+            for i in subset_indices:
+                metric_value = self._validation_fn(self._model, self._validation_dataset.get_data(indices=[i]))
```

Comment:
This part requires optimization, for example, we can spawn several processes, make model copies in each, allocate `per_sample_metrics` object in advance, and use nonoverlapping indices in each process.